File: graph/DFS/iterative/python/DFS.py
from collections.abc import Sequence


def DFS(graph: Sequence[Sequence[int]], start: int) -> list[int]:
    order: list[int] = []
    visited = [False] * len(graph)
    stack = [start]

    while stack:
        node = stack.pop()

        if visited[node]:
            continue

        visited[node] = True
        order.append(node)

        for neighbor in reversed(graph[node]):
            if not visited[neighbor]:
                stack.append(neighbor)

    return order

File: graph/DFS/iterative/rust/src/DFS.rs
#[allow(non_snake_case)]
pub fn DFS(graph: &[Vec<usize>], start: usize) -> Vec<usize> {
    let mut order = Vec::new();
    let mut visited = vec![false; graph.len()];
    let mut stack = vec![start];

    while let Some(node) = stack.pop() {
        if visited[node] {
            continue;
        }

        visited[node] = true;
        order.push(node);

        for &neighbor in graph[node].iter().rev() {
            if !visited[neighbor] {
                stack.push(neighbor);
            }
        }
    }

    order
}
